src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalMultiphasePoromechanics.hpp

```diff
@@ -0,0 +1,145 @@
+{
+
+/**
+ * @brief ThermalMultiphasePoromechanics strategy.
+ *
+ * Labels description stored in point_marker_array
+ *   - dofLabel: 0             = nodal displacement, x-component
+ *   - dofLabel: 1             = nodal displacement, y-component
+ *   - dofLabel: 2             = nodal displacement, z-component
+ *   - dofLabel: 3             = pressure
+ *   - dofLabel: 4             = density
+ *             ...             = densities
+ *   - dofLabel: numLabels - 2 = density
+ *   - dofLabel: numLabels - 1 = temperature
+ *
+ * 3-level MGR reduction strategy based on ThermalCompositionalMultiphaseFVM
+ *   - 1st level: eliminate displacements (0,1,2)
+ *   - 2nd level: eliminate the reservoir density associated with the volume constraint (numLabels - 1)
+ *   - 3nd level: eliminate the other reservoir densities
+ *   - The coarse grid (pressure and temperature) is solved with BoomerAMG.
+ *
+ */
+class ThermalMultiphasePoromechanics : public MGRStrategyBase< 3 >
+{
+public:
+  /**
+   * @brief Constructor.
+   * @param numComponentsPerField array with number of components for each field
+   */
+  explicit ThermalMultiphasePoromechanics( arrayView1d< int const > const & numComponentsPerField )
+    : MGRStrategyBase( LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[0] + numComponentsPerField[1] ) )
+  {
+    // Level 0: eliminate displacement degrees of freedom
+    m_labels[0].resize( m_numBlocks - 3 );
+    std::iota( m_labels[0].begin(), m_labels[0].end(), 3 );
+    // Level 1: eliminate last density which corresponds to the volume constraint equation
+    m_labels[1].resize( m_numBlocks - 5 );
+    std::iota( m_labels[1].begin(), m_labels[1].end(), 3 );
+    m_labels[1].push_back( m_numBlocks-1 ); // temperature
+    // Level 2: eliminate the remaining reservoir densities
+    m_labels[2].push_back( 3 ); // pressure
+    m_labels[2].push_back( m_numBlocks-1 ); // temperature
+
+    setupLabels();
+
+    // Level 0
+    m_levelFRelaxMethod[0]     = MGRFRelaxationMethod::amgVCycle;
+    m_levelInterpType[0]       = MGRInterpolationType::jacobi;
+    m_levelRestrictType[0]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[0] = MGRCoarseGridMethod::nonGalerkin;
+
+    // Level 1
+    m_levelFRelaxMethod[1]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[1]       = MGRInterpolationType::jacobi;
+    m_levelRestrictType[1]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[1] = MGRCoarseGridMethod::galerkin;
+
+    // Level 2
+    m_levelFRelaxMethod[2]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[2]       = MGRInterpolationType::injection;
+    m_levelRestrictType[2]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[2] = MGRCoarseGridMethod::cprLikeBlockDiag;
+
+    // ILU smoothing for the system made of pressure and densities
+    m_levelSmoothType[1]  = 16;
+    m_levelSmoothIters[1] = 1;
```

Comment:
Yes, that is something that I need to test again for the poromechanics case. We had looked into that with @MatthiasCremon for the flow-only case and had determined that this was the best option, but let's confirm this again for the poromechanics case

Reply:
Unrelated to MGR performance. I wonder if the [scaling](https://github.com/GEOSX/GEOS/blob/933cdc977ede8e18ed12bc2f9278832ed43e2ba4/src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp#L747-L753) used for the saturation-based volume balance play a role in the nonlinear convergence behavior. In other words, I wonder if the changes introduced in #2352 apply also to multiphase flow solvers in case of a non-zero rock compressibility.

Reply:
Yes. I am going to post the comparison results on this thread when the simulations are completed today or tomorrow

Reply:
@castelletto1 The full set of results for the isothermal case:

**Option 1:**

```
// ILU smoothing for the system made of pressure and densities
m_levelSmoothType[1]  = 16;
m_levelSmoothIters[1] = 1;
// Block GS smoothing for the system made of pressure and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations: 47.38028169
Avg. setup time (s): 5.08983618
Avg. solve time (s): 7.33402459

Number of time steps: 92                                                                                                                                            
Number of successful nonlinear iterations: 710                                                                                                                      
Number of successful linear iterations: 33640                                                                                                                       
Number of time step cuts: 0                                                                                                                                         
                                                                                                                                    
Total time                      9756.875s                                                                                                                                                
Initialization time              165.464s                                                                                                                                                
Run time                        9585.929s 

**Option 2:**

```
// ILU smoothing for the system made of pressure and densities (except the last one)
m_levelSmoothType[2]  = 16;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations: 50.15960452
Avg. setup time (s): 6.82931936
Avg. solve time (s): 8.02970959

Number of time steps: 92                                                                                                                                            
Number of successful nonlinear iterations: 708                                                                                                                      
Number of successful linear iterations: 35513                                                                                                                       
Number of time step cuts: 0                                                                                                                                         
                                                                                                                                     
Total time                     11472.665s                                                                                                                                                
Initialization time              165.359s                                                                                                                                                
Run time                       11298.852s  

**Option 3:**

```
// Block GS smoothing for the system made of pressure and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  51.60851064                                                                                                                                        
Avg. setup time (s):  4.7689113                                                                                                                                                        
Avg. solve time (s):  7.29502495

Number of time steps: 93                                                                                                                                            
Number of successful nonlinear iterations: 705                                                                                                                      
Number of successful linear iterations: 36384                                                                                                                       
Number of time step cuts: 0                                                                                                                                         
                                                                                                                                 
Total time                      9444.591s                                                                                                                                                
Initialization time              165.518s                                                                                                                                                
Run time                        9268.037s  

Almost there for the thermal case

Reply:
Also, I should say that this does not include the post-injection period, which is where the linear solver was struggling without ILU0. I wish I could run it quickly to show the results, but I burnt all my priority on Quartz. Anyway, more results will come over the next days

Reply:
I'm surprised that Option 3, compared to option 1, only needs on average just 3/4 iterations more per linear solve. It'd be interesting to have more granular timings for the linear solver performance. Something to be discussed during one of our next meetings with the hypre team.

Reply:
The thermal results:

**Option 1:**

```
// ILU smoothing for the system made of pressure, temperature and densities
m_levelSmoothType[1]  = 16;
m_levelSmoothIters[1] = 1;
// Block GS smoothing for the system made of pressure, temperature and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  34.43527204502814                                                                                                                                     
Avg. setup time (s):  8.569375909943716                                                                                                                                                  
Avg. solve time (s):  9.927093294559098  

Number of time steps: 100                                                                                                                                           
Number of successful nonlinear iterations: 533                                                                                                                      
Number of successful linear iterations: 18354                                                                                                                       
                                                                                                                           
Total time                     10488.448s                                                                                                                                                
Initialization time              175.963s                                                                                                                                                
Run time                       10305.489s 

**Option 2:**

```
// ILU smoothing for the system made of pressure, temperature and densities (except the last one)
m_levelSmoothType[2]  = 16;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  46.107074569789674                                                                                                                                    
Avg. setup time (s):  14.08313671128107                                                                                                                                                  
Avg. solve time (s):  13.30503089292543 

Number of time steps: 98                                                                                                                                            
Number of successful nonlinear iterations: 523                                                                                                                      
Number of successful linear iterations: 24114                                                                                                                       
                                                                                                                                    
Total time                     14990.070s                                                                                                                                                
Initialization time              193.887s                                                                                                                                                
Run time                       14783.487s 

**Option 3:**

```
// Block GS smoothing for the system made of pressure, temperature and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  157.1048951048951                                                                                                                                     
Avg. setup time (s):  7.8168362004661995                                                                                                                                                 
Avg. solve time (s):  40.64598799300698   

Does not complete before the time limit

Reply:
Thanks for running everything @francoishamon!
I'm surprised that Option 2 takes almost 2x time to setup the preconditioner compared to Option 1, when there is _less_ work to do (right?)... Am I missing an obvious reason?

Reply:
Yes, I agree, there are a few things that we do not understand well here. I will relay the questions to hypre

Reply:
The simulations took a while to start, but I finally have the results for the full isothermal simulation (injection period of 25 years, post-injection period of 200 years). I compare Option 1 and Option 3 to evaluate the impact of ILU:

**Option 1**:

```
// ILU smoothing for the system made of pressure and densities
m_levelSmoothType[1]  = 16;
m_levelSmoothIters[1] = 1;
// Block GS smoothing for the system made of pressure and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  48.10666667
Avg. setup time (s):  5.32633636
Avg. solve time (s):  7.552643

Number of time steps: 346                                                                                                                                           
Number of successful nonlinear iterations: 3000                                                                                                                     
Number of successful linear iterations: 144320                                                                                                                      
Number of time step cuts: 0                                                                                                                                         
Number of discarded nonlinear iterations: 0                                                                                                                         
Number of discarded linear iterations: 0                                                                                                                            
                                                                                                                                      
Total time                     42060.328s                                                                                                                                                
Initialization time              165.724s                                                                                                                                                
Run time                       41889.176s 

**Option 3**:

```
// Block GS smoothing for the system made of pressure and densities (except the last one)
m_levelSmoothType[2]  = 1;
m_levelSmoothIters[2] = 1;
```

Avg. number of linear iterations:  64.84694866                                                                                                                                         
Avg. setup time (s):  5.12443929                                                                                                                                                       
Avg. solve time (s):  9.41923532   

Number of time steps: 353                                                                                                                                           
Number of successful nonlinear iterations: 3047                                                                                                                     
Number of successful linear iterations: 199024                                                                                                                      
Number of time step cuts: 1                                                                                                                                         
Number of discarded nonlinear iterations: 50                                                                                                                        
Number of discarded linear iterations: 1807    
                                                                                                                                                                                                                                                            
Total time                     48606.235s                                                                                                                                                
Initialization time              168.248s                                                                                                                                                
Run time                       48431.114s